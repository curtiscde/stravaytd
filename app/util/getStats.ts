import fs from 'fs';
import { IYtdHistory } from '../types/IYtdHistory';
import { addFinalDateToYtdHistory } from './addFinalDateToYtdHistory';
import { filterData } from './filterData';
import { generateRealTimeYtd } from './generateRealTimeYtd';
import { getAthletesCurrentYtd } from './getAthletesCurrentYtd';
import { mapAthleteName } from './mapAthleteName';

const convertMtoKm = (metres: number) => Math.round((metres / 1000) * 100) / 100;

const getStyles = (index: number) => {
  const borderColors = [
    'rgba(255, 99, 132, 1)',
    'rgba(54, 162, 235, 1)',
    'rgba(255, 206, 86, 1)',
    'rgba(75, 192, 192, 1)',
    'rgba(153, 102, 255, 1)',
    'rgba(255, 159, 64, 1)',
  ];
  const backgroundColors = [
    'rgba(255, 99, 132, 0.2)',
    'rgba(54, 162, 235, 0.2)',
    'rgba(255, 206, 86, 0.2)',
    'rgba(75, 192, 192, 0.2)',
    'rgba(153, 102, 255, 0.2)',
    'rgba(255, 159, 64, 0.2)',
  ];
  return {
    borderColor: borderColors[index],
    backgroundColor: backgroundColors[index],
  };
};

interface GetStats {
  allowedAthletes: string
  year: number
}

export function getStats({ allowedAthletes, year }: GetStats) {
  const ytdHistoryFile = fs.readFileSync('../app/data/ytdHistory.json', 'utf-8');
  const ytd = JSON.parse(ytdHistoryFile) as IYtdHistory;
  const currentYtds = getAthletesCurrentYtd('./data/current-ytd');
  const now = new Date().getTime();
  let realTimeYtd = generateRealTimeYtd({ ytd, currentYtds, now });
  realTimeYtd = addFinalDateToYtdHistory({ ytdHistory: realTimeYtd });

  const { athletes }: any = realTimeYtd;

  const distanceData = {
    datasets: filterData({
      athletes,
      propertyName: 'distance',
      year,
    }).map((athlete: any, index: any) => ({
      label: mapAthleteName(allowedAthletes, athlete.athleteId),
      ...getStyles(index),
      data: athlete.ytd.map((data: any) => ({
        x: data.date, y: convertMtoKm(data.distance),
      })),
    })),
  };

  const runsData = {
    datasets: filterData({
      athletes,
      propertyName: 'distance',
      year,
    }).map((athlete: any, index: any) => ({
      label: mapAthleteName(allowedAthletes, athlete.athleteId),
      ...getStyles(index),
      data: athlete.ytd.map((data: any) => ({
        x: data.date, y: data.count,
      })),
    })),
  };

  const movingTimeData = {
    datasets: filterData({
      athletes,
      propertyName: 'distance',
      year,
    }).map((athlete: any, index: any) => ({
      label: mapAthleteName(allowedAthletes, athlete.athleteId),
      ...getStyles(index),
      data: athlete.ytd.map((data: any) => ({
        x: data.date, y: (data.movingTime / 60),
      })),
    })),
  };

  const elevationGainData = {
    datasets: filterData({
      athletes,
      propertyName: 'distance',
      year,
    }).map((athlete: any, index: any) => ({
      label: mapAthleteName(allowedAthletes, athlete.athleteId),
      ...getStyles(index),
      data: athlete.ytd.map((data: any) => ({
        x: data.date, y: data.elevationGain,
      })),
    })),
  };

  return {
    distanceData,
    runsData,
    movingTimeData,
    elevationGainData,
  };
}
