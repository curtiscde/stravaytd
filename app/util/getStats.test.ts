import { getStats } from './getStats';

const convertToTimestamp = (date: string) => new Date(date).getTime();

jest.mock('fs', () => ({
  readFileSync: () => JSON.stringify({
    athletes: [
      {
        athleteId: 123,
        ytd: [
          {
            date: convertToTimestamp('2021-02-02'), distance: 123, movingTime: 456, elevationGain: 7, count: 8,
          },
          {
            date: convertToTimestamp('2022-01-01'), distance: 10000, movingTime: 300, elevationGain: 3, count: 4,
          },
          {
            date: convertToTimestamp('2022-03-01'), distance: 12000, movingTime: 600, elevationGain: 7, count: 8,
          },
        ],
      },
      {
        athleteId: 456,
        ytd: [
          {
            date: convertToTimestamp('2021-02-02'), distance: 123, movingTime: 456, elevationGain: 7, count: 8,
          },
          {
            date: convertToTimestamp('2022-01-01'), distance: 20000, movingTime: 300, elevationGain: 4, count: 5,
          },
          {
            date: convertToTimestamp('2022-03-01'), distance: 80000, movingTime: 900, elevationGain: 10, count: 11,
          },
        ],
      },
    ],
  }),
}));
jest.mock('./getAthletesCurrentYtd', () => ({
  getAthletesCurrentYtd: () => [],
}));

describe('getStats', () => {
  let stats: any;

  describe('year 2022', () => {
    beforeAll(() => {
      stats = getStats({ allowedAthletes: '123:foo,456:bar', year: 2022 });
    });

    it('returns stats', () => {
      expect(stats).toMatchSnapshot();
    });
  });
});
