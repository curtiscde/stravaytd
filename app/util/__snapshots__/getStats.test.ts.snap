// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStats year 2022 returns stats 1`] = `
Object {
  "distanceData": Object {
    "datasets": Array [
      Object {
        "backgroundColor": "rgba(255, 99, 132, 0.2)",
        "borderColor": "rgba(255, 99, 132, 1)",
        "data": Array [
          Object {
            "x": 1640995200000,
            "y": 10,
          },
          Object {
            "x": 1646092800000,
            "y": 12,
          },
        ],
        "label": "foo",
      },
      Object {
        "backgroundColor": "rgba(54, 162, 235, 0.2)",
        "borderColor": "rgba(54, 162, 235, 1)",
        "data": Array [
          Object {
            "x": 1640995200000,
            "y": 20,
          },
          Object {
            "x": 1646092800000,
            "y": 80,
          },
        ],
        "label": "bar",
      },
    ],
  },
  "elevationGainData": Object {
    "datasets": Array [
      Object {
        "backgroundColor": "rgba(255, 99, 132, 0.2)",
        "borderColor": "rgba(255, 99, 132, 1)",
        "data": Array [
          Object {
            "x": 1640995200000,
            "y": 3,
          },
          Object {
            "x": 1646092800000,
            "y": 7,
          },
        ],
        "label": "foo",
      },
      Object {
        "backgroundColor": "rgba(54, 162, 235, 0.2)",
        "borderColor": "rgba(54, 162, 235, 1)",
        "data": Array [
          Object {
            "x": 1640995200000,
            "y": 4,
          },
          Object {
            "x": 1646092800000,
            "y": 10,
          },
        ],
        "label": "bar",
      },
    ],
  },
  "movingTimeData": Object {
    "datasets": Array [
      Object {
        "backgroundColor": "rgba(255, 99, 132, 0.2)",
        "borderColor": "rgba(255, 99, 132, 1)",
        "data": Array [
          Object {
            "x": 1640995200000,
            "y": 5,
          },
          Object {
            "x": 1646092800000,
            "y": 10,
          },
        ],
        "label": "foo",
      },
      Object {
        "backgroundColor": "rgba(54, 162, 235, 0.2)",
        "borderColor": "rgba(54, 162, 235, 1)",
        "data": Array [
          Object {
            "x": 1640995200000,
            "y": 5,
          },
          Object {
            "x": 1646092800000,
            "y": 15,
          },
        ],
        "label": "bar",
      },
    ],
  },
  "runsData": Object {
    "datasets": Array [
      Object {
        "backgroundColor": "rgba(255, 99, 132, 0.2)",
        "borderColor": "rgba(255, 99, 132, 1)",
        "data": Array [
          Object {
            "x": 1640995200000,
            "y": 4,
          },
          Object {
            "x": 1646092800000,
            "y": 8,
          },
        ],
        "label": "foo",
      },
      Object {
        "backgroundColor": "rgba(54, 162, 235, 0.2)",
        "borderColor": "rgba(54, 162, 235, 1)",
        "data": Array [
          Object {
            "x": 1640995200000,
            "y": 5,
          },
          Object {
            "x": 1646092800000,
            "y": 11,
          },
        ],
        "label": "bar",
      },
    ],
  },
}
`;
